mitodl/ol-keycloak
# 1882: Keycloak display identity providers linked with the user on the login password page

# What are the relevant tickets?
https://github.com/mitodl/ol-infrastructure/issues/1882

# Description (What does it do?)
A. Identifies which IDp's a user is linked with on the login-password page when the username of the user is known.  The username of the user can be known when the browser-login flow contains the Username Form before the Open Learning Username Password Form, for example:
![Screenshot 2023-11-06 at 11 08 38 AM](https://github.com/mitodl/ol-keycloak/assets/8311573/ebd9e5d1-e755-43fe-8288-d52260c077ef)

B. Adds in the custom Open Learning Keycloak theme that used to exist here (https://github.com/mitodl/ol-keycloak-customization).  The theme is included into this repo so that custom extensions and the theme are compiled into jar files at the same time.  This will hopefully make managing and deploying Keycloak customizations easier and only require a single repository for both theme and extensions.



# Screenshots (if appropriate):
**View for a user linked with a single IDP**
![Screenshot 2023-11-06 at 11 04 43 AM](https://github.com/mitodl/ol-keycloak/assets/8311573/8ccb3845-96c1-4b92-a0ad-d8be6bb1593e)

**View for a user not linked with any IDP**
![Screenshot 2023-11-06 at 11 07 48 AM](https://github.com/mitodl/ol-keycloak/assets/8311573/19c48ac8-aa0c-41ed-b9c4-eb984863dd66)


# How can this be tested?

1. Follows the steps outlined in the PR's README in order to integrate the PR changes into a local instance of Keycloak.
2. Create 2 dummy Identity Providers. (details don't matter, just ensure the names are different).
3. Create a user and manually link them to one of the identity providers from step 2 using the Keycloak Admin Console.  You can enter anything into the IDP link form.
4. Set the login theme to use the custom OL theme included in this PR.
5. Create a browser login flow that matches screenshot in the "Description" section of this PR.
6. Logout of Keycloak and attempt to log back in using the user from step 3.
7. Verify that the username-password page displays the IDP which the user is connected with and the IDP the user is not connect with (should look like the "View for a user linked with a single IDP" screenshot).
8. Log back into the Keycloak admin console.
9. Link the user from step 3 to both IDPs from step 2.
10. Perform step 6.
11. Verify that the username-password page displays the IDP which the user is connected with.


# Notes
This PR does not include unit tests.  Keycloak does not provide an easy API to build unit tests on top of.  Because of that, there will need to be considerable work done to build infrastructure into this repository in order to support unit tests.  I will create an issue for this work to be done in the future.
